extensions/parquet/base/src/main/java/io/deephaven/parquet/base/util/SeekableChannelsProviderLoader.java

```diff
@@ -0,0 +1,59 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.parquet.base.util;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ServiceLoader;
+
+/**
+ * A service loader class for loading {@link SeekableChannelsProviderPlugin} implementations at runtime and provide
+ * {@link SeekableChannelsProvider} implementations for different URI schemes, e.g., S3.
+ */
+public final class SeekableChannelsProviderLoader {
+
+    private static volatile SeekableChannelsProviderLoader instance;
+
+    public static SeekableChannelsProviderLoader getInstance() {
+        if (instance == null) {
+            instance = new SeekableChannelsProviderLoader();
+        }
+        return instance;
+    }
+
+    private final List<SeekableChannelsProviderPlugin> providers;
+
+    private SeekableChannelsProviderLoader() {
+        providers = new ArrayList<>();
+    }
+
+    /**
+     * Create a new {@link SeekableChannelsProvider} based on given URI and object using the plugins loaded by the
+     * {@link ServiceLoader}. For example, for a "S3" URI, we will create a {@link SeekableChannelsProvider} which can
+     * read files from S3.
+     *
+     * @param uri The URI
+     * @param object An optional object to pass to the {@link SeekableChannelsProviderPlugin} implementations.
+     * @return A {@link SeekableChannelsProvider} for the given URI.
+     */
+    public SeekableChannelsProvider fromServiceLoader(@NotNull final URI uri, @Nullable final Object object) {
```

Comment:
There's a concurrency issue here. Multiple calls to `fromServiceLoader` could stomp on each other while reading or updating `providers`.

One reasonable approach might be to load the plugins into an umodifiable list at construction time, unless we need to preserve the notion that there might be alternative configuration in the future. Alternatively, this code needs to introduce synchronization and safety around `providers`.